libp2p/Host.cpp

```diff
@@ -759,36 +746,50 @@ void Host::startedWorking()
         m_run = true;
     }
 
-    // start capability threads (ready for incoming connections)
-    for (auto const& h: m_capabilities)
-        h.second->onStarting();
-    
-    // try to open acceptor (todo: ipv6)
-    int port = Network::tcp4Listen(m_tcp4Acceptor, m_netConfig);
-    if (port > 0)
+    if (m_capabilities.size())
     {
-        m_listenPort = port;
-        determinePublic();
-        runAcceptor();
+        // start capability threads (ready for incoming connections)
+        for (auto const& h : m_capabilities)
+            h.second->onStarting();
+
+        // try to open acceptor (todo: ipv6)
+        int port = Network::tcp4Listen(m_tcp4Acceptor, m_netConfig);
+        if (port > 0)
+        {
+            // Port may have been changed if the specified port was already in use
+            m_netConfig.listenPort = m_listenPort = port;
+            determinePublic();
+            runAcceptor();
+        }
+        else
+            LOG(m_logger) << "p2p.start.notice id: " << id()
+                          << " TCP Listen port is invalid or unavailable.";
     }
-    else
-        LOG(m_logger) << "p2p.start.notice id: " << id() << " TCP Listen port is invalid or unavailable.";
-
-    auto nodeTable = make_shared<NodeTable>(
-        m_ioService,
-        m_alias,
-        NodeIPEndpoint(bi::address::from_string(listenAddress()), listenPort(), listenPort()),
-        m_netConfig.discovery,
-        m_netConfig.allowLocalDiscovery
-    );
-    nodeTable->setEventHandler(new HostNodeTableHandler(*this));
+
+    auto nodeTable = make_shared<NodeTable>(m_ioService, m_alias,
+        // Use data from network configuration rather than Host's settings because p2p might not
+        // have been started
+        NodeIPEndpoint(!m_netConfig.listenIPAddress.empty() ?
+                           bi::address::from_string(m_netConfig.listenIPAddress) :
+                           bi::address_v4(),
+            m_netConfig.listenPort /* UDP */, m_netConfig.listenPort /* TCP */),
+        m_netConfig.discovery, m_netConfig.allowLocalDiscovery);
+
+    if (m_capabilities.size())
+        // Only set the event handler if there are capabilities since otherwise p2p
+        // isn't running
+        nodeTable->setEventHandler(new HostNodeTableHandler(*this));
     DEV_GUARDED(x_nodeTable)
-        m_nodeTable = nodeTable;
+    m_nodeTable = nodeTable;
     restoreNetwork(&m_restoreNetwork);
 
-    LOG(m_logger) << "p2p.started id: " << id();
-
-    run(boost::system::error_code());
+    if (m_capabilities.size())
+    {
+        LOG(m_logger) << "p2p.started id: " << id();
```

Comment:
This displays our NodeID, so might be useful for the bootnode, too. On the other hand, there's another log message somewhere also printing it at program start, so maybe one of them should be removed.

Reply:
Correct - the host's enode info is already logged in both aleth's and aleth-bootnode's main. I think we should still log a status message regarding the started/not started p2p status of aleth/aleth-bootnode but I don't think we need to log the ID along with the status. I'll remove the id() call.

Reply:
The ID is also logged in the Host ctor: https://github.com/ethereum/aleth/blob/91787f239566501a2b23f17946618ddf6a3cee15/libp2p/Host.cpp#L93-L95

I'll remove it from there as well.

Reply:
I've actually ended up moving the enode URL logging code from the mains to right after network start in Host::start - this consolidates things slightly.